### DISCUSSION TITLE
Understanding complex API rules

### DISCUSSION THREAD

**Discussion by tonnax** [pocketbase/pocketbase#4062] [Q&A] [answered]
<img width="1494" alt="image" src="https://github.com/pocketbase/pocketbase/assets/140943390/f7c66698-b56d-4184-839d-ba0488db3a52">

How can I create an API rule for UPDATE on "likes" collection where:
The request.auth.id must be a user that belong to the group (see user_groups collection for user and group relation) where the post this user's like is updated. 

TLDR: I don't want the user that is not member of group A can like the post that belong to group A.

**Comment by ganigeorgiev** [comment, marked as answer]
I'm not sure why you may need so many intertwined references to the "users" collection, but if I'm understanding correctly what you are trying to do, then the following rule should work:

```js
@request.auth.id != "" &&
users_id.id ?= @request.auth.id &&
// back relation references are not supported yet, so we have to "join" them manually
@collection.user_groups.users_id.id ?= @request.auth.id &&
@collection.user_groups.group_id.id ?= post_id.group_id
```

Also note that having the `likes.users_id` as multiple relation field will allow any user that can create/update likes to modify the liked "users_id" array. If this is not intended, I would recommend to use a more traditional SQL structure and change the field to a single relation (eg. `user_id`) and define a UNIQUE index constraint from the Admin UI for the "user_id-post_id" pair to disallow a user to "like" twice the same post.

**Reply by tonnax** [reply]
Awesome man, I thought it would be a very long syntax but it is simple enough for me to dig deeper.
I also just migrated my 'likes' collection users_id from multi to single as you suggested, thank you!

Question: how can I prevent user to "like" twice the same post using Pocketbase admin API syntax? (rules for CREATE)

**Reply by ganigeorgiev** [reply]
> how can I prevent user to "like" twice the same post using Pocketbase admin API syntax?

I don't recommend using API rules for that. Instead define a combined `UNIQUE` index from the _Admin UI > {collection} > Edit collection > "Unique constraints and indexes"_ section.
![unique_index_popup](https://i.imgur.com/xLvlQjp.png)

**Reply by tonnax** [reply]
Wow, magic! Where can I learn more about this feature @ganigeorgiev? I didn't see it on the documentation. 
So I assume this feature is **solely** for helping to prevent duplicate record inside a collection. Thank you once again!

**Reply by ganigeorgiev** [reply]
There is no dedicated "Unique constraints and indexes" section in the PocketBase documentation but this feature is based on the SQLite indexes implementation so you can read more in:
- Basic syntax info: https://www.sqlite.org/lang_createindex.html
- Partial indexes (aka. indexes with a `WHERE` clause): https://www.sqlite.org/partialindex.html

**Reply by tonnax** [reply]
Gotcha! Back to your answer question Gani
```
> @collection.user_groups.users_id.id
```
that field is a multi relation to users table. I wonder what will that syntax will return? will it return list/array of id of users_id? or is it returning the first id of users_id?

**Reply by ganigeorgiev** [reply]
It doesn't return anything. It will produce an operand that will apply a condition on the `id` field of the `users_id` relations.

Whether the condition will be match-all or any/at-least-one-of depends on the operator - `=` (match all) or `?=` (at-least-one-of).

If in the above we used `=` it will require **every** `users_id` relation record from **every** `user_groups` collection record to have and `id` that equals to the right side.

You can explore the generated SQL queries in the console with the `--dev` flag when starting the executable.
